File: 프로그래머스/3/49191. 순위/순위.java
class Solution {
    public int solution(int n, int[][] results) {
        int answer = 0;
        int[][] ranks = new int[n][n];
        
        for (int i = 0; i < results.length; i++) {
            int[] rank = results[i];
            ranks[rank[0]-1][rank[1]-1] = 1;
            ranks[rank[1]-1][rank[0]-1] = -1;
        }
        
        
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (i == j) continue;
                    
                    if (ranks[i][j] != 0) continue;
                    
                    if (ranks[i][k] == 1 && ranks[k][j] == 1) {
                        ranks[i][j] = 1;
                        ranks[j][i] = -1;
                    }
                }
            }
        }
        for (int i = 0; i < n; i++) {
            int cnt = 0;
            for (int j = 0; j < n; j++) {
                if (ranks[i][j] == 0) cnt++;
            }
            
            if (cnt == 1) answer++;
        }
        
        return answer;
    }
}

File: 프로그래머스/3/49191. 순위/순위.py
def solution(n, results):
    answer = 0
    
    rank = [[0] * n for _ in range(n)]
    
    for win, lose in results:
        rank[win-1][lose-1] = 1
        rank[lose-1][win-1] = -1
        
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if i == j:
                    continue
                if rank[i][j] != 0:
                    continue # 이미 연결 되어 있는 상태니깐
                    
                if rank[i][k] == 1 and rank[k][j] == 1:
                    rank[i][j] = 1
                    rank[j][i] = -1
                    
    for per in rank:
        if per.count(0) == 1:
            answer += 1
    
    return answer